### ISSUE TITLE
Opflow takes the adjoint of an Operator if the expectation value is written as ``OperatorMeasurement``

### ISSUE THREAD

**Issue by Cryoris** [Qiskit/qiskit-terra#6254]
### Information

- **Qiskit Terra version**: master @  1c049978525ea912422fb656625264aa7d4075bd
- **Python version**: 3.7.9
- **Operating system**: macOS Big Sur

### What is the current behavior?

As discovered by @Zoufalc, if an expectation value is written as `~StateFn(operator) @ state`, the operator is complex conjugated although it shouldn't be. I.e. in that case the Opflow evaluates 
```
<state | operator_dagger | state>
```
instead of the expected 
```
<state | operator | state>
```

### Steps to reproduce the problem

```python
from qiskit.opflow import *
op = 1j * Z
state = Zero
print((~Zero @ op @ Zero).eval()) # = 1j -> correct
print((~StateFn(op) @ Zero).eval()) # = -1j -> wrong sign! The operator was complex conjugated!
```

### What is the expected behavior?

Don't take the adjoint of the operator.

### Suggested solutions

Maybe we should adjust `OperatorStateFn.adjoint` to not take the adjoint of the primitive? Not sure if that breaks stuff elsewhere.

**Comment by ikkoham** [comment]
I thought this behavior is as expected. `~StateFn(operator) @ state` means Hilbert-Schmidt inner product <operator | "state">_HS = tr[operator^dagger |state><state|].

**Comment by Cryoris** [comment]
I don't think so, as far as I remember this is supposed to evaluate the expectation value of the operator as it is. 

Since several expectation converters only support this representation of `~StateFn(op) @ state` it would be strange if that wouldn't be an ordinary expectation. I think e.g. VQE uses this representation, or the tutorials. 

Maybe we should have an explicit state function for measurements to avoid this confusion?

**Comment by Zoufalc** [comment]
If the `operator` is Hermitian then this is always fine since then `operator^dagger = operator`.  However, this is not true in general and then we need to be careful and ensure that
`<state | operator | state>  = tr[operator |state><state|]`.

**Comment by ikkoham** [comment]
Evaluation of the expectation value of the operator is given by Hilbert-Schmidt inner product. That is natural from the perspective of mathematics and this is used in PTM representation.
I know this representation is frequently used.

As @Zoufalc said, if the operator is Hermitian, there are no problem. For Non-Hermitian operator,
` <state | operator | state> =  tr[operator |state><state|] = <operator^\dagger | "state">_HS` where `"state" =  |state><state|`.
What is important here is whether the representation `~StateFn(operator) @ state` means `<operator^\dagger | "state">_HS` or `tr[operator |state><state|]`.
I strongly support the representation means the HS inner product because ~ represents adjoint.
If you want to calculate `<state | operator | state>`, just write `print((~StateFn(~op) @ Zero).eval())`

> Maybe we should have an explicit state function for measurements to avoid this confusion?

This may be nice idea!

**Comment by ikkoham** [comment]
The [opflow tutorial](https://qiskit.org/documentation/tutorials/operators/01_operator_flow.html) said

> The composition operator is used to perform an inner product, which by default is held in an unevaluated form. 

So this representation must be inner product.

**Comment by Cryoris** [comment]
For the case of the left operand being a `OperatorStateFn` we have a special case, see https://qiskit.org/documentation/tutorials/operators/01_operator_flow.html#OperatorStateFn.

**Comment by Zoufalc** [comment]
I think we need to be careful here since the Hilbert-Schmidt inner product is -as you say- defined as `<A|B>=Tr(A^*B)` but an expectation value w.r.t. a state `|s>` and observable `O` is given by `Tr(O)|s><s| = <s|O|s>`.

**Comment by Cryoris** [comment]
tl;dr: We should keep the current behavior but add a class `OperatorMeasurement` to explicitly do expectation values.

Ok so there's an abuse of notation we need to clear up. Here's what I think:

* The canonical way to write expectation values in Opflow is `OperatorMeasurement(operator) @ state`. I say canonical because expectation converters and gradients work only with that representation and not with `~state @ operator @ state`.
* We've been using `~StateFn(op)` as shorthand notation to obtain a `OperatorMeasurement(operator)`. But that's not technically correct, and this is where @ikkoham is right: `~` means adjoint and means we complex conjugate the state function. What we _meant_ with `~StateFn(op) @ state` is _actually_ `StateFn(op, is_measurement=True)`. If you run the above example, this works correctly:
    ```
    print((StateFn(op, is_measurement=True) @ Zero).eval()) # = 1j -> correct
    ```
* To clear this confusion up we need a designated way to write operator measurements. Since the adjoint of an operator state function prints as "OperatorMeasurement" I suggest we add a class `OperatorMeasurement` to represent the adjoint of `OperatorStateFn`. With that the syntax would be
    ```
    print((OperatorMeasurement(op) @ Zero).eval()) # = 1j -> correct
    ```

**Comment by ikkoham** [comment]
@Cryoris Thank you for your nice suggestion. I completely agree with your idea.

**Comment by jlapeyre** [comment]
If we want to fix this, I think we should take the opportunity to make the design a bit simpler and more uniform. It may be possible to do this without too much disruption. I'd like to

* Separate the concepts of measurement and adjoint, which are currently confounded. We could do this by leaving the current design mostly intact, but adding a new way to represent what we have been calling "expectation" or "measurement".

* Have the compose operator `@`, which for the most part constructs `ComposedOp`, represent one idea; probably tensor contraction. It seems this already agrees with how it is used in opflow, except for things like `StateFn(op, is_measurement=True) @ Zero`, in which `@` means something different. If we want `@` to mean tensor contraction, then we could represent the latter as  `~state @ op @ state`. Or if that is difficult to recognize or handle, then something like `DiagMatrixElement(op, state)`, which does not abuse the meaning of `@`.

* Use technical quantum mechanics terms only in their usual textbook sense; roughly Nielsen and Chuang. In particular, "measurement" in QM never means "a complex linear combination of expectation values of observables", which is, it seems, how we have been using it.

* In particular, I disagree with  using `OperatorMeasurement(op) @ Zero` on two counts: 1) `@` is used anomalously and confusingly here. and 2) `OperatorMeasurement(op)` seems to suggest a QM measurement. But, really it means, *treat the following symbol `@` in a special way, ie construct a diagonal matrix element <0| op |0>, where `op` is an arbitrary matrix.*

**Comment by Cryoris** [comment]
Closing this, as from above it turns out this is not a bug in the code but rather a bug in my user expectation 😉
